src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -129,58 +135,87 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
         // GetAssetAsync call will most likely cache hit. it is most likely since we might change cache heuristic in
         // future which make data to live a lot shorter in the cache, and the data might get expired before one actually
         // consume the data. 
-        using (Logger.LogBlock(FunctionId.AssetService_SynchronizeProjectAssetsAsync, Checksum.GetProjectChecksumsLogInfo, projectChecksums, cancellationToken))
+        using (Logger.LogBlock(FunctionId.AssetService_SynchronizeProjectAssetsAsync, message: null, cancellationToken))
         {
             await SynchronizeProjectAssetsWorkerAsync().ConfigureAwait(false);
         }
 
         async ValueTask SynchronizeProjectAssetsWorkerAsync()
         {
-            // get children of project checksum objects at once
+            using var _ = ArrayBuilder<Task>.GetInstance(out var tasks);
+
+            // Make parallel requests for all the project data across all projects at once.
+            tasks.Add(SynchronizeProjectAssetAsync<ProjectInfo.ProjectAttributes>(static p => p.Info, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetAsync<CompilationOptions>(static p => p.CompilationOptions, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetAsync<ParseOptions>(static p => p.ParseOptions, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetCollectionAsync<ProjectReference>(static p => p.ProjectReferences, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetCollectionAsync<MetadataReference>(static p => p.MetadataReferences, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetCollectionAsync<AnalyzerReference>(static p => p.AnalyzerReferences, cancellationToken));
+
+            // Then sync each project's documents in parallel with each other.
+            foreach (var projectChecksums in allProjectChecksums)
+                tasks.Add(SynchronizeProjectDocumentsAsync(projectChecksums, cancellationToken));
```

Comment:
>cancellationToken [](http://example.com/codeflow?start=77&length=17)

Any concern about a potential large increase in cancellation exceptions?

Reply:
Can cross that bridge if it happens :-)